llvm/lib/Passes/PassBuilderPipelines.cpp

```diff
@@ -1293,6 +1298,21 @@ PassBuilder::buildModuleOptimizationPipeline(OptimizationLevel Level,
   if (RunPartialInlining)
     MPM.addPass(PartialInlinerPass());
 
+  // Scheduling LoopVersioningLICM when inlining is over, because after that
+  // we may see more accurate aliasing. Reason to run this late is that too
+  // early versioning may prevent further inlining due to increase of code
+  // size. By placing it just after inlining other optimizations which runs
+  // later might get benefit of no-alias assumption in clone loop.
+  if (UseLoopVersioningLICM) {
+    MPM.addPass(createModuleToFunctionPassAdaptor(
+        createFunctionToLoopPassAdaptor(LoopVersioningLICMPass())));
+    MPM.addPass(
+        createModuleToFunctionPassAdaptor(createFunctionToLoopPassAdaptor(
+            LICMPass(PTO.LicmMssaOptCap, PTO.LicmMssaNoAccForPromotionCap,
+                     /*AllowSpeculation=*/true),
+            /*USeMemorySSA=*/true, /*UseBlockFrequencyInfo=*/false)));
+  }
```

Comment:
I think this should be moved into the OptimizePM below. There doesn't seems to be a good reason to run this earlier with a separate FPM.

Reply:
It's ok and I have moved it into the OptimizePM. Thanks for your advice.

Reply:
Why did you only move the LICM pass, and not the LoopVersioningLICM pass as well?

Reply:
My point is that LoopVersioningLICM pass is better to be as early as possible to reuse the following optimizations because the copy codes will be added no alias metadata. and it can be with LICM pass together certainly, what's your opinion about it? thanks

Reply:
I don't see any passes that could benefit from LoopVersioningLICM before OptimizePM.

Reply:
That's ok and I have moved LoopVersioningLICM into OptimizePM, and LoopVersioningLICM and LICM pass use  each FunctionToLoopPassAdaptor  with different parameters.